src/components/vm/disks/diskAdd.jsx

```diff
@@ -365,104 +361,92 @@ const CustomPath = ({ idPrefix, onValueChanged, device, validationFailed, hideDe
-        const availableTarget = getNextAvailableTarget(existingTargets, defaultBus);
-        const sortFunction = (poolA, poolB) => poolA.name.localeCompare(poolB.name);
-        let defaultPool;
-        if (storagePools.length > 0)
-            defaultPool = storagePools
-                    .map(pool => ({ name: pool.name, type: pool.type }))
-                    .sort(sortFunction)[0];
-
-        return {
-            storagePools,
-            vm,
-            vms,
-            file: "",
-            device: "disk",
-            storagePoolName: defaultPool && defaultPool.name,
-            storagePoolType: defaultPool && defaultPool.type,
-            mode: isMediaInsertion ? CUSTOM_PATH : CREATE_NEW,
-            volumeName: "",
-            existingVolumeName: undefined,
-            size: 1,
-            unit: units.GiB.name,
-            format: defaultPool && getDefaultVolumeFormat(defaultPool),
-            target: availableTarget,
-            permanent: !domainIsRunning(vm.state), // default true for a down VM; for a running domain, the disk is attached tentatively only
-            hotplug: domainIsRunning(vm.state), // must be kept false for a down VM; the value is not being changed by user
-            addDiskInProgress: false,
-            cacheMode: 'default',
-            busType: defaultBus,
-            updateDisks: false,
-            serial: "",
-        };
-    }
+    const [diskParams, setDiskParams] = useState(initialDiskParams);
+    const storagePoolName = diskParams.storagePool?.name;
+    const storagePoolType = diskParams.storagePool?.type;
+
+    const Dialogs = useDialogs();
 
-    componentDidMount() {
+    useEffect(() => {
         // Refresh storage volume list before displaying the dialog.
         // There are recently no Libvirt events for storage volumes and polling is ugly.
         // https://bugzilla.redhat.com/show_bug.cgi?id=1578836
-        storagePoolGetAll({ connectionName: this.props.vm.connectionName })
-                .then(() => this.setState({ dialogLoading: false, ...this.initialState }))
+        storagePoolGetAll({ connectionName: vm.connectionName })
+                .always(() => setDialogLoading(false))
                 .catch(exc => this.dialogErrorSet(_("Storage pools could not be fetched"), exc.message));
-    }
+    }, [vm.connectionName]);
 
-    validateParams() {
+    const validateParams = () => {
         const validationFailed = {};
 
-        if (this.state.mode !== CUSTOM_PATH && !this.state.storagePoolName)
+        if (mode !== CUSTOM_PATH && !storagePoolName)
             validationFailed.storagePool = _("Please choose a storage pool");
 
-        if (this.state.mode === CUSTOM_PATH && this.state.customDiskVerificationFailed)
-            validationFailed.customPath = this.state.customDiskVerificationMessage;
+        if (mode === CUSTOM_PATH && customDiskVerificationFailed)
+            validationFailed.customPath = customDiskVerificationMessage;
```

Comment:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1303-20231109-142627-1ac1b404-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)